### PR TITLE
Upgrade to TypeScript v2.1.4

### LINKED ISSUE #1572: TSLint errors when checking async fat arrow functions because of missing parentheses
### Bug Report
- **TSLint version**: `3.15.0-dev.0`
- **TypeScript version**: `2.1.0-dev.20160923`
- **Running TSLint via**: CLI
#### TypeScript code being linted

``` ts
private attachEventListeners ()
    {
        $("table#testTrackFilterSync td.autoSync").on ("contextmenu", async (event) =>
        {
            // some code..
        });
    }
```

with `tslint.json` configuration:

``` json

{
 "rules" :   {
        "member-access" : true
        , "no-inferrable-types": true
        , "ban": 
            [
                true
                , 
                [
                    "console", "functionTrace"
                ]
            ]
        , "curly" : true
        , "no-console" : [true, "debug"]
        , "no-empty"   : true
        , "no-eval"    : true
        , "no-unreachable": true
        , "eofline": true
        , "indent": [true, "spaces"]
        , "no-trailing-whitespace": true

        , "align": [true, "parameters", "statements", "arguments"]
        , "arrow-parens": true
        , "class-name" : true
        , "new-parens" : true
        , "no-consecutive-blank-lines": true
        , "one-line": [false, "check-catch", "check-finally", "check-else", "check-whitespace", "check-open-brace"]
        , "quotemark": [true, "double"]
        , "semicolon": [true, "always"]
        , "variable-name": [true, "ban-keywords", "check-format", "allow-leading-underscore"]
        , "whitespace": [true, "check-branch", "check-operator", "check-typecast", "check-decl", "check-type"]

    }
}
```
#### Actual behavior

TSlint gives me an error that I should write parentheses around the parameters
Without the async modifier it works fine.
#### Expected behavior

No error should be given

### PR DESCRIPTION
**Pull request by nchen63**
Lots of stuff broken by the TS API

fixes #1572

### CODE REVIEW

File: src/rules/restrictPlusOperandsRule.ts [hunk 1]
```diff
@@ -47,22 +44,28 @@ class RestrictPlusOperandsWalker extends Lint.ProgramAwareRuleWalker {
     public visitBinaryExpression(node: ts.BinaryExpression) {
         if (node.operatorToken.kind === ts.SyntaxKind.PlusToken) {
             const tc = this.getTypeChecker();
-            const leftType = tc.typeToString(tc.getTypeAtLocation(node.left));
-            const rightType = tc.typeToString(tc.getTypeAtLocation(node.right));
+
+            const leftType = getAddableType(node.left, tc);
```
Comment (nchen63) [inline src/rules/restrictPlusOperandsRule.ts]: had to rewrite this because tc.typeToString() doesn't work

**Comment by JKillian** [comment]
If you haven't, might be worth making sure all the tests still pass when using TS 2.x, since users might have a peer dependency on that version.

**Comment by JKillian** [comment]
Also, this might fix https://github.com/palantir/tslint/issues/1458

**Comment by nchen63** [comment]
Do you mean writing version specific code? Like:

```
if (typescript.version === 2.0.x) {
    do something
} else if (typescript.version === 2.1.x) {
     do something else
}
```

**Comment by JKillian** [comment]
More likely some sort of feature detection perhaps?

```
if (ts.oldRemovedFunc != null) {
  ts.oldRemovedFunc()
} else {
  ts.newCorrectFunc()
}
```

File: src/rules/noConditionalAssignmentRule.ts [hunk 2]
```diff
@@ -60,7 +60,7 @@ class NoConditionalAssignmentWalker extends Lint.RuleWalker {
 
     protected visitDoStatement(node: ts.DoStatement) {
         this.validateConditionalExpression(node.expression);
-        super.visitWhileStatement(node);
+        super.visitDoStatement(node);
```
Comment (adidahiya) [inline src/rules/noConditionalAssignmentRule.ts]: nice catch 👍

**Pull request merged**